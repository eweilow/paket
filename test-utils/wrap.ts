import { ChildProcess, spawn } from "child_process";
import { join } from "path";

export function wrapCommand<T extends { [key: string]: string }>(cmd: string, cwd: string, env: T) {
  let childProcess: ChildProcess;

  async function start() {
    const nodeExec = process.env.npm_node_execpath || process.execPath;
    childProcess = spawn(nodeExec, [join(__dirname, "../packages/cli/cli.js"), ...cmd.split(" ")], {
      cwd,
      env
    });
  }

  function stopped() {
    return new Promise(resolve => {
      if (childProcess == null) {
        resolve();
      }

      childProcess.on("exit", () => {
        childProcess = null;
        resolve();
      });
    });
  }

  async function stop() {
    if (childProcess == null) {
      return;
    }
    const onStopped = stopped();
    childProcess.kill();
    await onStopped;
  }

  type PipeFn<TPipe> = (msg: TPipe) => void;

  let currentOutFn: PipeFn<any>;
  let currentErrFn: PipeFn<any>;
  return {
    start,
    stop,
    async wait(forStr: string) {
      if (childProcess == null) {
        throw new Error("childProcess is null!");
      }
      return new Promise((resolve, reject) => {
        const listener = data => {
          if (data.toString().indexOf(forStr) >= 0) {
            childProcess.stdout.off("data", listener);
            resolve();
          }
        };
        childProcess.stdout.on("data", listener);
        childProcess.on("exit", code => {
          if (code !== 0) {
            reject("Exited with code " + code);
          }
        });
      });
    },
    pipe(out: PipeFn<string>, err: PipeFn<string>) {
      if (childProcess == null) {
        throw new Error("childProcess is null!");
      }

      currentOutFn = data => out(data.toString());
      currentErrFn = data => err(data.toString());
      childProcess.stdout.on("data", currentOutFn);
      childProcess.stderr.on("data", currentErrFn);
    },
    unpipe() {
      childProcess.stdout.off("data", currentOutFn);
      childProcess.stderr.off("data", currentErrFn);
    },
    stopped
  };
}
