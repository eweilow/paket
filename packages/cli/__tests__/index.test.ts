import express from "express";
import { promises } from "fs";
import { Server } from "http";
import { join } from "path";
import { DirectoryResult, dir } from "tmp-promise";

import { wrapCommand } from "../../../test-utils/wrap";

describe("CLI", () => {
  let server: Server;
  beforeAll(() => {
    const app = express();
    app.use(express.static(join(__dirname, "./version-mocks"), { index: "index.json" }));
    server = app.listen(1234);
  });

  afterAll(cb => {
    server.close(cb);
  });

  let tmpDir: DirectoryResult;
  beforeEach(async () => {
    tmpDir = await dir({ unsafeCleanup: true });
  });

  afterEach(async () => {
    tmpDir.cleanup();
  });

  function stringifyAndSplit(data: string[], ...filters: string[]) {
    const msgs = data
      .join("\n")
      .split(/\n|\r\n/g)
      .map(el => el.trim())
      .filter(el => !!el)
      .filter(el => !filters.find(filter => el.includes(filter)));
    return msgs;
  }

  async function run(cmd: string) {
    const start = wrapCommand(cmd, tmpDir.path, {
      PAKET_REGISTRY: "http://localhost:1234"
    });

    await start.start();
    const out: string[] = [];
    const err: string[] = [];
    start.pipe(
      m => !m.includes("Using root folder") && out.push(m),
      m => !m.includes("Using root folder") && err.push(m)
    );
    await start.wait("Using registry");
    await start.stopped();

    expect(stringifyAndSplit(out)).toMatchSnapshot("stdout");
    expect(stringifyAndSplit(err)).toMatchSnapshot("stderr");
  }

  describe("uses any mode", () => {
    it("runs correctly with no package.json", async () => {
      await run("update any @types/node @types/jest");
    });

    it("runs correctly with empty package.json", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: {},
            devDependencies: {},
            optionalDependencies: {},
            resolutions: {},
            resolveModules: []
          },
          null,
          "  "
        )
      );
      await run("update any @types/node @types/jest");
    });

    it("runs correctly with dependencies in package.json", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: { "@types/node": "1.0.0", "@types/jest": "^1.0.0" },
            devDependencies: { "@types/node": "0.1.0", "@types/jest": "^5.0.0" },
            optionalDependencies: { "@types/node": "1.1.0", "@types/jest": "^1.0.5" },
            resolutions: { "**/@types/node": "1.0.1", "**/@types/jest": "1.0.3" },
            resolveModules: []
          },
          null,
          "  "
        )
      );
      await run("update any @types/node @types/jest");
    });

    it("runs resolutions correctly", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: { "@types/node": "1.0.0", "@types/jest": "^1.0.0" },
            devDependencies: { "@types/node": "0.1.0", "@types/jest": "^5.0.0" },
            optionalDependencies: { "@types/node": "1.1.0", "@types/jest": "^1.0.5" },
            resolutions: { "**/@types/node": "1.0.1", "**/@types/jest": "^1.0.3" },
            resolveModules: ["@types/node"]
          },
          null,
          "  "
        )
      );
      await run("update any @types/node @types/jest");
    });
  });

  describe("uses latest mode", () => {
    it("runs correctly with no package.json", async () => {
      await run("update latest @types/node @types/jest");
    });

    it("runs correctly with empty package.json", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: {},
            devDependencies: {},
            optionalDependencies: {},
            resolutions: {},
            resolveModules: []
          },
          null,
          "  "
        )
      );
      await run("update latest @types/node @types/jest");
    });

    it("runs correctly with dependencies in package.json", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: { "@types/node": "1.0.0", "@types/jest": "^1.0.0" },
            devDependencies: { "@types/node": "0.1.0", "@types/jest": "^5.0.0" },
            optionalDependencies: { "@types/node": "1.1.0", "@types/jest": "^1.0.5" },
            resolutions: { "**/@types/node": "1.0.1", "**/@types/jest": "1.0.3" },
            resolveModules: []
          },
          null,
          "  "
        )
      );
      await run("update latest @types/node @types/jest");
    });

    it("runs resolutions correctly", async () => {
      await promises.writeFile(
        join(tmpDir.path, "./package.json"),
        JSON.stringify(
          {
            name: "pkg-name",
            dependencies: { "@types/node": "1.0.0", "@types/jest": "^1.0.0" },
            devDependencies: { "@types/node": "0.1.0", "@types/jest": "^5.0.0" },
            optionalDependencies: { "@types/node": "1.1.0", "@types/jest": "^1.0.5" },
            resolutions: { "**/@types/node": "1.0.1", "**/@types/jest": "^1.0.3" },
            resolveModules: ["@types/node"]
          },
          null,
          "  "
        )
      );
      await run("update latest @types/node @types/jest");
    });
  });
});
