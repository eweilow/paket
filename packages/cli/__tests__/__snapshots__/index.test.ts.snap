// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI uses any mode runs correctly with dependencies in package.json: stderr 1`] = `Array []`;

exports[`CLI uses any mode runs correctly with dependencies in package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the last published version.",
  "pkg-name:",
  "@types/node (normal): 1.0.0 -> ^3.0.0",
  "@types/jest (normal): ^1.0.0 -> ^3.0.0",
  "@types/node (dev): 0.1.0 -> ^3.0.0",
  "@types/jest (dev): ^5.0.0 -> ^3.0.0",
  "@types/node (optional): 1.1.0 -> ^3.0.0",
  "@types/jest (optional): ^1.0.5 -> ^3.0.0",
]
`;

exports[`CLI uses any mode runs correctly with empty package.json: stderr 1`] = `Array []`;

exports[`CLI uses any mode runs correctly with empty package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the last published version.",
  "No updates required.",
]
`;

exports[`CLI uses any mode runs correctly with no package.json: stderr 1`] = `Array []`;

exports[`CLI uses any mode runs correctly with no package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the last published version.",
  "No updates required.",
]
`;

exports[`CLI uses any mode runs resolutions correctly: stderr 1`] = `Array []`;

exports[`CLI uses any mode runs resolutions correctly: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the last published version.",
  "pkg-name:",
  "@types/node (normal): 1.0.0 -> ^3.0.0",
  "@types/jest (normal): ^1.0.0 -> ^3.0.0",
  "@types/node (dev): 0.1.0 -> ^3.0.0",
  "@types/jest (dev): ^5.0.0 -> ^3.0.0",
  "@types/node (optional): 1.1.0 -> ^3.0.0",
  "@types/jest (optional): ^1.0.5 -> ^3.0.0",
  "@types/node (resolutions): 1.0.1 -> 3.0.0",
]
`;

exports[`CLI uses latest mode runs correctly with dependencies in package.json: stderr 1`] = `Array []`;

exports[`CLI uses latest mode runs correctly with dependencies in package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the latest version.",
  "pkg-name:",
  "@types/node (normal): 1.0.0 -> ^2.0.0",
  "@types/jest (normal): ^1.0.0 -> ^3.0.0",
  "@types/node (dev): 0.1.0 -> ^2.0.0",
  "@types/jest (dev): ^5.0.0 -> ^3.0.0",
  "@types/node (optional): 1.1.0 -> ^2.0.0",
  "@types/jest (optional): ^1.0.5 -> ^3.0.0",
]
`;

exports[`CLI uses latest mode runs correctly with empty package.json: stderr 1`] = `Array []`;

exports[`CLI uses latest mode runs correctly with empty package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the latest version.",
  "No updates required.",
]
`;

exports[`CLI uses latest mode runs correctly with no package.json: stderr 1`] = `Array []`;

exports[`CLI uses latest mode runs correctly with no package.json: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the latest version.",
  "No updates required.",
]
`;

exports[`CLI uses latest mode runs resolutions correctly: stderr 1`] = `Array []`;

exports[`CLI uses latest mode runs resolutions correctly: stdout 1`] = `
Array [
  "Using registry 'http://localhost:1234'",
  "Updating packages matching globs:",
  "- @types/node",
  "- @types/jest",
  "to the latest version.",
  "pkg-name:",
  "@types/node (normal): 1.0.0 -> ^2.0.0",
  "@types/jest (normal): ^1.0.0 -> ^3.0.0",
  "@types/node (dev): 0.1.0 -> ^2.0.0",
  "@types/jest (dev): ^5.0.0 -> ^3.0.0",
  "@types/node (optional): 1.1.0 -> ^2.0.0",
  "@types/jest (optional): ^1.0.5 -> ^3.0.0",
  "@types/node (resolutions): 1.0.1 -> 2.0.0",
]
`;
